syntax = "proto3";

package airsim_bridge;

service DroneControl {
  // Bidirectional stream for observations and actions
  rpc ControlLoop(stream DroneAction) returns (stream DroneObservation) {}

  // Request a reset of the environment
  rpc ResetEnvironment(ResetRequest) returns (DroneObservation) {}
}

message Vector3 {
  float x = 1;
  float y = 2;
  float z = 3;
}

message Quaternion {
  float w = 1;
  float x = 2;
  float y = 3;
  float z = 4;
}

message Image {
  int32 width = 1;
  int32 height = 2;
  bytes data = 3;
  string format = 4; // "rgb", "depth", etc.
}

message DroneObservation {
  Vector3 position = 1;
  Vector3 linear_velocity = 2;
  Vector3 angular_velocity = 3;
  Quaternion orientation = 4;
  repeated Image images = 5;
  float collision_info = 6;
  float reward = 7;
  bool done = 8;
  int64 timestamp_ns = 9;
}

message DroneAction {
  float roll = 1;         // -1 to 1
  float pitch = 2;        // -1 to 1
  float yaw_rate = 3;     // -1 to 1
  float throttle = 4;     // 0 to 1
  int64 timestamp_ns = 5;
}

message ResetRequest {
  bool random_start = 1;
  Vector3 initial_position = 2;
  Quaternion initial_orientation = 3;
}
